src/platform/audio/ffmpegaudiofilecache.cpp

```diff
@@ -0,0 +1,449 @@
+        Close_FFmpeg_Contexts(file);
+        return false;
+    }
+
+    ret = avformat_find_stream_info(file->fmt_ctx, NULL);
+    if (ret < 0) {
+        captainslog_error("Failed to find stream info");
+        Close_FFmpeg_Contexts(file);
+        return false;
+    }
+
+    if (file->fmt_ctx->nb_streams != 1) {
+        captainslog_error("Expected exactly one audio stream per file");
+        Close_FFmpeg_Contexts(file);
+        return false;
+    }
+
+    AVCodec *input_codec = avcodec_find_decoder(file->fmt_ctx->streams[0]->codecpar->codec_id);
+    if (!input_codec) {
+        captainslog_error("Audio codec not supported: '%u'", file->fmt_ctx->streams[0]->codecpar->codec_tag);
+        Close_FFmpeg_Contexts(file);
+        return false;
+    }
+
+    ret = avcodec_open2(file->fmt_ctx->streams[0]->codec, input_codec, NULL);
+    if (ret < 0) {
+        captainslog_error("Failed to open input codec");
+        Close_FFmpeg_Contexts(file);
+        return false;
+    }
+
+    file->codec_ctx = file->fmt_ctx->streams[0]->codec;
+    return true;
+}
+
+/**
+ * Decode the input data and append it to our wave data stream
+ */
+bool FFmpegAudioFileCache::Decode_FFmpeg(FFmpegOpenAudioFile *file)
+{
+    AVPacket *packet = av_packet_alloc();
+    AVFrame *frame = av_frame_alloc();
+
+    int result = 0;
+    while (av_read_frame(file->fmt_ctx, packet) >= 0) {
+        result = avcodec_send_packet(file->codec_ctx, packet);
+        if (result < 0) {
+            captainslog_error("Failed to send audio packet to decoder.");
+            return false;
+        }
+        result = avcodec_receive_frame(file->codec_ctx, frame);
+        // Check if this was a real error or we just need more data
+        if (result < 0 && result != AVERROR(EAGAIN) && result != AVERROR_EOF) {
+            captainslog_error("Failed to receive audio frame from decoder.");
+            return false;
+        } else if (result >= 0) {
+            int frame_data_size = av_samples_get_buffer_size(
+                NULL, file->codec_ctx->channels, frame->nb_samples, file->codec_ctx->sample_fmt, 1);
+            av_realloc(file->wave_data, file->data_size + frame_data_size);
+            memcpy(file->wave_data + file->data_size, frame->data[0], frame_data_size);
+            file->data_size += frame_data_size;
+        }
+        av_packet_unref(packet);
+    }
+
+    av_packet_free(&packet);
+    av_frame_free(&frame);
+
+    return true;
+}
+
+/**
+ * Close all the open FFmpeg handles for an open file.
+ */
+void FFmpegAudioFileCache::Close_FFmpeg_Contexts(FFmpegOpenAudioFile *file)
+{
+    if (file->fmt_ctx) {
+        avformat_close_input(&file->fmt_ctx);
+    }
+
+    if (file->avio_ctx) {
+        av_freep(&file->avio_ctx);
+    }
+}
+
+void FFmpegAudioFileCache::Fill_Wave_Data(FFmpegOpenAudioFile *file)
+{
+    WavHeader wav;
+    wav.chunk_size = file->data_size - 8;
+    wav.subchunk2_size = file->data_size - 44;
```

Comment:
What do these magic number here mean?

Reply:
Those are the the offsets of the fields inside the WAV struct. Will use `offsetof`